src/dbnode/persist/fs/msgpack/encoder_fast.go

```diff
@@ -0,0 +1,348 @@
+		buf[5] = byte(v >> 24)
+		buf[6] = byte(v >> 16)
+		buf[7] = byte(v >> 8)
+		buf[8] = byte(v)
+		return b
+	}
+
+	if v >= int64(int8(codes.NegFixedNumLow)) {
+		b, buff := growAndReturn(b, 1)
+		buff[0] = byte(v)
+		return b
+	}
+
+	if v >= math.MinInt8 {
+		// Equivalent to: return write1(b, codes.Int8, uint64(v))
+		b, buf := growAndReturn(b, 2)
+		buf[0] = codes.Int8
+		buf[1] = byte(uint64(v))
+		return b
+	}
+
+	if v >= math.MinInt16 {
+		// Equivalent to: return write2(b, codes.Int16, uint64(v))
+		b, buf := growAndReturn(b, 3)
+		n := uint64(v)
+		buf[0] = codes.Int16
+		buf[1] = byte(uint64(n) >> 8)
+		buf[2] = byte(uint64(n))
+		return b
+	}
+
+	if v >= math.MinInt32 {
+		// Equivalent to: return write4(b, codes.Int32, uint64(v))
+		b, buf := growAndReturn(b, 5)
+		n := uint64(v)
+		buf[0] = codes.Int32
+		buf[1] = byte(n >> 24)
+		buf[2] = byte(n >> 16)
+		buf[3] = byte(n >> 8)
+		buf[4] = byte(n)
+		return b
+	}
+
+	// Equivalent to: return write8(b, codes.Int64, uint64(v))
+	b, buf := growAndReturn(b, 9)
+	n := uint64(v)
+	buf[0] = codes.Int64
+	buf[1] = byte(n >> 56)
+	buf[2] = byte(n >> 48)
+	buf[3] = byte(n >> 40)
+	buf[4] = byte(n >> 32)
+	buf[5] = byte(n >> 24)
+	buf[6] = byte(n >> 16)
+	buf[7] = byte(n >> 8)
+	buf[8] = byte(n)
+	return b
+}
+
+// encodeFloat64 is used in EncodeLogEntryFast which is a very hot path.
+// As a result, many of the function calls in this function have been
+// manually inlined to reduce function call overhead.
+func encodeFloat64(b []byte, v float64) []byte {
+	// Equivalent to: return write8(b, codes.Double, math.Float64bits(n))
+	b, buf := growAndReturn(b, 9)
+	buf[0] = codes.Double
+	n := math.Float64bits(v)
+	buf[1] = byte(n >> 56)
+	buf[2] = byte(n >> 48)
+	buf[3] = byte(n >> 40)
+	buf[4] = byte(n >> 32)
+	buf[5] = byte(n >> 24)
+	buf[6] = byte(n >> 16)
+	buf[7] = byte(n >> 8)
+	buf[8] = byte(n)
+	return b
+}
+
+// encodeBytes is used in EncodeLogEntryFast which is a very hot path.
+// As a result, many of the function calls in this function have been
+// manually inlined to reduce function call overhead.
+func encodeBytes(b []byte, data []byte) []byte {
+	if data == nil {
+		b, buf := growAndReturn(b, 1)
+		buf[0] = codes.Nil
+		return b
+	}
+
+	// Equivalent to: encodeBytesLen(b, len(data))
+	var (
+		l   = len(data)
+		v   = uint64(l)
+		buf []byte
+	)
+	if l < 256 {
+		b, buf = growAndReturn(b, 2)
+		buf[0] = codes.Bin8
+		buf[1] = byte(uint64(l))
+	} else if l < 65536 {
+		b, buf = growAndReturn(b, 3)
+		buf[0] = codes.Bin16
+		buf[1] = byte(v >> 8)
+		buf[2] = byte(v)
+		return b
+	} else {
+		b, buf = growAndReturn(b, 5)
+		buf[0] = codes.Bin32
+		buf[1] = byte(v >> 24)
+		buf[2] = byte(v >> 16)
+		buf[3] = byte(v >> 8)
+		buf[4] = byte(v)
+	}
+
+	b = append(b, data...)
+	return b
+}
+
+// encodeBytesLen is not used, it is left here to demonstrate what a manually inlined
+// version of this function should look like.
+func encodeBytesLen(b []byte, l int) []byte {
```

Comment:
Is this called? Looks Inlined elsewhere.  This is where codgen would help (some being reused while others being used as real calls, everything can just always be inlined potentially).

Reply:
Nope not called (if it is, it’s because I forgot to in-line it). I just left these in as “documention” per the comment